### DISCUSSION TITLE
How to show a default value in ItemInput?

### DISCUSSION THREAD

**Discussion by jiriks74** [forntoh/LcdMenu#56] [Q&A] [answered]
So I have a settings menu with some values I have as default or that I load from eeprom.
I have an `ItemInput("Text", callaback)` and I'd like to show some numbers as default (eg timings for display brightness). The current behaviour is that there's nothing (`Text:`) and I'd like it to show the value that's already set (`Text:2000`)

Any ideas?

**Comment by forntoh** [comment, marked as answer]
`ItemInput` has two constructors, you can add a default value `:String` using the first constructor

```cpp
ItemInput(const char* text, String value, fptrStr callback)
    : MenuItem(text, value, callback, MENU_ITEM_INPUT) {}

ItemInput(const char* text, fptrStr callback)
    : MenuItem(text, "", callback, MENU_ITEM_INPUT) {}
```

**Reply by timetofly06** [reply]
Hi !
Great pleasure to discover your library. It works like a charm but I have some difficulties to apply this method. 

Using the first constructor with "string value" variable seems to be a one shot initialization. When I change the value of the variable in the void setup, value stay the same

String texte = "5555";
   MenuItem mainMenu[] = {ItemHeader(),
                       ItemInput("Con", texte, inputCallback),
   ....

In your tutorial examples, "ItemInput" is declared before the "Void Setup" of the sketch
eeprom.read command must appear on the void/loop section. 

 When I try to move the ItemInput in the void section, I have some compilation errors saying the 'menu' object is not defined in the loop, of course.

I search for hours and hours but I have some difficulties to find a workaround.  
I found this topic (https://github.com/forntoh/LcdMenu/discussions/9

So I applied it  ( "1" to take the first item) :

menu[1]->value = "xxxx" combined with menu.enter() and menu.back() 
but it is not very easy to use when you have several variables to initialise dispatched in few submenus. 

Maybe someone can help me to find a better solution ? Thanks for all guys

**Reply by forntoh** [reply]
I had to use `String` since it has awesome functions for string manipulation needed for the `ItemInput`. I'm currently working on making it a `char * ` **maybe** it would help you since the value will effectively become a pointer to some part of memory and you can easily change where it points to.

**Reply by timetofly06** [reply]
Thanks for your quick answer 👍  And, yeah, it could be great. For now, I applied the following workaround :

In the first part of the code, I described the mainmenu item and all the Submenu description containing ItemSubmenu with default value on second parameter of the ItemInput.

Then, on the void setup section, I check the EEPROM content. If it contents my software name in the first adresses used as a tag, I will just have to overwrite the defaults parameters set on the above part of code.  I proceed like this for the first submenu parameters 

    menu.enter();
    menu[1]->value = String_param1;
    menu[2]->value = String_param2;
    menu[3]->value = String_param3;
    menu.back();
    menu.back();

  menu.down()   // to go to the second submenu

  .... and so on

Checking the EEprom, when the tag is not found, I set the tag and all other parameters with eeprom.put method

The only problem ( but it is not a real one) if that all submenus are quickly scrolled in order to assign values read from the EEPROM when you power on or reset.

By the way, as the string lenght is not limited by a user-defined format, we can get some "bonus chars" during the input of the value. It could be awesome to add a parameter format on the ItemInput keyword to control it. (for example : XXXXX for a 5 chars lenght or 9.9 or 99.9 for a decimal value). Do you think it could be possible ? I can post it ont eh ideas sections ;-)

**Reply by forntoh** [reply]
You can simply do

```cpp
menu[1]->value = String_param1;
menu[2]->value = String_param2;
menu[3]->value = String_param3;
menu.resetMenu();
```

There's a reset method `resetMenu` which repaints the display, added in v2.1.3 #79 
Just call it after changing the values

**Reply by forntoh** [reply]
> By the way, as the string lenght is not limited by a user-defined format, we can get some "bonus chars" during the input of the value. It could be awesome to add a parameter format on the ItemInput keyword to control it. (for example : XXXXX for a 5 chars lenght or 9.9 or 99.9 for a decimal value). Do you think it could be possible ? I can post it ont eh ideas sections ;-)

I don't really understand what you mean, go ahead and post it in the ideas section with more context.

**Reply by timetofly06** [reply]
> You can simply do
> 
> ```c++
> menu[1]->value = String_param1;
> menu[2]->value = String_param2;
> menu[3]->value = String_param3;
> menu.resetMenu();
> ```
> 
> There's a reset method `resetMenu` which repaints the display, added in v2.1.3 #79 Just call it after changing the values

Sorry for my late reply, I was out of my home and thank you so much for your help. I just tried the resetMenu method, but I can't make it works.

//    menu.enter();
    menu[1]->value = sCHF_Seuil;
    menu[2]->value = sCHF_Duree;
    menu[3]->value = sCHF_Ecart0;
    menu[4]->value = sCHF_Ecart1;
    menu[5]->value = sCHF_Nbre;
    menu.resetMenu();
    
//    menu.back();
//    menu.back();

First I checked in the lcdmenu.h and I found the "void resetMenu()", so I called it  like that but unfortunately, values are not updated
I tried with 2.1.3 librairies downloaded from IDE and also directly from Github (Date of files librairies are older from IDE)

I noticed that resetMenu() stays in black color in the editor. I think the command is not interpreted.  I also just have one library at a time in Windows\user\documents\Arduino\Libraries

![image](https://user-images.githubusercontent.com/60317152/236339781-044e87ac-9470-4375-a0ad-d54a5c98e3e2.png)


IT WORKS ! I am replying to myself : My parameters are set on few submenus.  It seems that we can't avoid to enter on each submenus to make them refreshed

**Reply by forntoh** [reply]
I have replaced String with char* in this PR #86
